nimare/meta/cbmr.py

```diff
@@ -0,0 +1,1827 @@
+                    self.inputs_["moderators_by_group"] = moderators_by_group
+            
+                foci_per_voxel, foci_per_study = dict(), dict()
+                for group in studies_by_group.keys():
+                    group_study_id = studies_by_group[group]
+                    group_coordinates = dataset.coordinates.loc[
+                        dataset.coordinates["study_id"].isin(group_study_id)
+                    ]
+                    # Group-wise foci coordinates
+                    # Calculate IJK matrix indices for target mask
+                    # Mask space is assumed to be the same as the Dataset's space
+                    group_xyz = group_coordinates[["x", "y", "z"]].values
+                    group_ijk = mm2vox(group_xyz, mask_img.affine)
+                    group_foci_per_voxel = np.zeros(mask_img.shape, dtype=np.int32)
+                    for ijk in group_ijk:
+                        group_foci_per_voxel[ijk[0], ijk[1], ijk[2]] += 1
+                    # will not work with maskers that aren't NiftiMaskers
+                    group_foci_per_voxel = nib.Nifti1Image(
+                        group_foci_per_voxel, mask_img.affine, mask_img.header
+                    )
+                    group_foci_per_voxel = masker.transform(group_foci_per_voxel).transpose()
+                    # number of foci per voxel/study
+                    n_group_study = len(group_study_id)
+                    group_foci_per_study = np.array(
+                        [(group_coordinates["study_id"] == i).sum() for i in group_study_id]
+                    )
+                    group_foci_per_study = group_foci_per_study.reshape((n_group_study, 1))
+
+                    foci_per_voxel[group] = group_foci_per_voxel
+                    foci_per_study[group] = group_foci_per_study
+
+                self.inputs_["foci_per_voxel"] = foci_per_voxel
+                self.inputs_["foci_per_study"] = foci_per_study
+
+    def _model_structure(self, model, penalty, device):
+        """Specify stochastic models for CBMR with or without Firth-type penalty.
+
+        For stochastic models, there're three options: Poisson, NB, clustered NB models.
+        For penalty term, we only consider Firth-type penalty currently.
+
+        Parameters
+        ----------
+        model : :obj:`str`
+            Name of stochastic model in CBMR: Poisson, NB or clustered NB models.
+        penalty : :obj:`bool`
+            Whether to penalize log-likelihood function with Firth-type penalty.
+        device : :obj:`str`
+            Device type ('cpu' or 'cuda') represents the device on which operations will
+            be allocated
+        """
+        beta_dim = self.inputs_["coef_spline_bases"].shape[1]  # regression coef of spatial effect
+        if self.moderators:
+            gamma_dim = list(self.inputs_["moderators_by_group"].values())[0].shape[1]
+            study_level_moderators = True
```

Comment:
this variable isn't necessary, gamma_dim (i.e., moderators) can be used instead.